build-system/babel-plugins/babel-plugin-config-transformer/index.js

```diff
@@ -0,0 +1,59 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+let hasThirdPartyRegex = false;
+let hasCdnProxyRegex = false;
+
+// Only executes on the ESM build, intended to use with the Google AMP Cache.
+module.exports = function({types: t}) {
+  return {
+    visitor: {
```

Comment:
We can just handle all of this with a MemberExpression: https://astexplorer.net/#/gist/1a179b93507ca03bbe286f34f36063f6/b779cd1ef2448c2ad4db1bdafae0f635b8e151cf

Reply:
When we do this, Closure doesn't remove everything as [expected](https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520SIMPLE_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%252F%252F%2520ADD%2520YOUR%2520CODE%2520HERE%250Aconst%2520thirdPartyFrameRegex%2520%253D%2520typeof%2520void%25200%2520%253D%253D%2520'string'%2520%253F%2520new%2520RegExp(void%25200)%2520%253A%2520void%25200%253B%250Aconst%2520cdnProxyRegex%2520%253D%2520typeof%2520void%25200%2520%253D%253D%2520'string'%2520%253F%2520new%2520RegExp(void%25200)%2520%253A%2520void%25200%253B%250Aconst%2520urls%2520%253D%2520%257B%250A%2520%2520thirdParty%253A%2520void%25200%2520%257C%257C%2520'https%253A%252F%252F3p.ampproject.net'%252C%250A%2520%2520thirdPartyFrameHost%253A%2520void%25200%2520%257C%257C%2520'ampproject.net'%252C%250A%2520%2520thirdPartyFrameRegex%253A%2520thirdPartyFrameRegex%2520%257C%257C%2520%252F%255Ed-%255Cd%252B%255C.ampproject%255C.net%2524%252F%252C%250A%2520%2520cdn%253A%2520void%25200%2520%257C%257C%2520'https%253A%252F%252Fcdn.ampproject.org'%252C%250A%2520%2520cdnProxyRegex%253A%2520cdnProxyRegex%2520%257C%257C%2520%252F%255Ehttps%253A%255C%252F%255C%252F(%255Ba-zA-Z0-9_-%255D%252B%255C.)%253Fcdn%255C.ampproject%255C.org%2524%252F%252C%250A%2520%2520localhostRegex%253A%2520%252F%255Ehttps%253F%253A%255C%252F%255C%252Flocalhost(%253A%255Cd%252B)%253F%2524%252F%252C%250A%2520%2520errorReporting%253A%2520void%25200%2520%257C%257C%2520'https%253A%252F%252Famp-error-reporting.appspot.com%252Fr'%252C%250A%2520%2520localDev%253A%2520void%25200%2520%257C%257C%2520false%252C%250A%2520%2520trustedViewerHosts%253A%2520%255B%252F(%255E%257C%255C.)google%255C.(com%253F%257C%255Ba-z%255D%257B2%257D%257Ccom%253F%255C.%255Ba-z%255D%257B2%257D%257Ccat)%2524%252F%252C%2520%252F(%255E%257C%255C.)gmail%255C.(com%257Cdev)%2524%252F%255D%252C%250A%2520%2520geoApi%253A%2520void%25200%250A%257D%250Aconsole.log(urls)).



```javascript
var thirdPartyFrameRegex=void 0,cdnProxyRegex=void 0,urls={thirdParty:"https://3p.ampproject.net",thirdPartyFrameHost:"ampproject.net",thirdPartyFrameRegex:thirdPartyFrameRegex||/^d-\d+\.ampproject\.net$/,cdn:"https://cdn.ampproject.org",cdnProxyRegex:cdnProxyRegex||/^https:\/\/([a-zA-Z0-9_-]+\.)?cdn\.ampproject\.org$/,localhostRegex:/^https?:\/\/localhost(:\d+)?$/,errorReporting:"https://amp-error-reporting.appspot.com/r",localDev:!1,trustedViewerHosts:[/(^|\.)google\.(com?|[a-z]{2}|com?\.[a-z]{2}|cat)$/,
/(^|\.)gmail\.(com|dev)$/],geoApi:void 0};console.log(urls);
```

Reply:
You've got to run it with advance transforms to enable the DCE stuff:
[test](https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%252F%252F%2520ADD%2520YOUR%2520CODE%2520HERE%250Aconst%2520env%2520%253D%2520self.AMP_CONFIG%2520%257C%257C%2520%257B%257D%253B%250Aconst%2520thirdPartyFrameRegex%2520%253D%2520typeof%2520void%25200%2520%253D%253D%2520'string'%2520%253F%2520new%2520RegExp(void%25200)%2520%253A%2520void%25200%253B%250Aconst%2520cdnProxyRegex%2520%253D%2520typeof%2520void%25200%2520%253D%253D%2520'string'%2520%253F%2520new%2520RegExp(void%25200)%2520%253A%2520void%25200%253B%250A%252F**%2520%2540type%2520%257B!Object%253Cstring%252C%2520string%257Cboolean%257CRegExp%257CArray%253CRegExp%253E%253E%257D%2520*%252F%250A%250Aexport%2520const%2520urls%2520%253D%2520%257B%250A%2520%2520thirdParty%253A%2520void%25200%2520%257C%257C%2520'https%253A%252F%252F3p.ampproject.net'%252C%250A%2520%2520thirdPartyFrameHost%253A%2520void%25200%2520%257C%257C%2520'ampproject.net'%252C%250A%2520%2520thirdPartyFrameRegex%253A%2520thirdPartyFrameRegex%2520%257C%257C%2520%252F%255Ed-%255Cd%252B%255C.ampproject%255C.net%2524%252F%252C%250A%2520%2520cdn%253A%2520void%25200%2520%257C%257C%2520'https%253A%252F%252Fcdn.ampproject.org'%252C%250A%2520%2520cdnProxyRegex%253A%2520cdnProxyRegex%2520%257C%257C%2520%252F%255Ehttps%253A%255C%252F%255C%252F(%255Ba-zA-Z0-9_-%255D%252B%255C.)%253Fcdn%255C.ampproject%255C.org%2524%252F%252C%250A%2520%2520localhostRegex%253A%2520%252F%255Ehttps%253F%253A%255C%252F%255C%252Flocalhost(%253A%255Cd%252B)%253F%2524%252F%252C%250A%2520%2520errorReporting%253A%2520void%25200%2520%257C%257C%2520'https%253A%252F%252Famp-error-reporting.appspot.com%252Fr'%252C%250A%2520%2520localDev%253A%2520void%25200%2520%257C%257C%2520false%252C%250A%2520%2520trustedViewerHosts%253A%2520%255B%252F(%255E%257C%255C.)google%255C.(com%253F%257C%255Ba-z%255D%257B2%257D%257Ccom%253F%255C.%255Ba-z%255D%257B2%257D%257Ccat)%2524%252F%252C%2520%252F(%255E%257C%255C.)gmail%255C.(com%257Cdev)%2524%252F%255D%252C%250A%2520%2520geoApi%253A%2520void%25200%250A%257D%253B%250Awindow.config%2520%253D%2520%257B%250A%2520%2520urls%250A%257D%253B)

Can you try running `gulp dist` to see if it's eliminated?

Reply:
Running it with `gulp dist` also didn't eliminate it.

Reply:
I check because if it didn't, that meant the compiler wasn't doing basic DCE which means we had bigger problems. But I got:

```js
  const Ka = {
        thirdParty: 'https://3p.ampproject.net',
        thirdPartyFrameHost: 'ampproject.net',
        thirdPartyFrameRegex: /^d-\d+\.ampproject\.net$/,
        cdn: 'https://cdn.ampproject.org',
        cdnProxyRegex: /^https:\/\/([a-zA-Z0-9_-]+\.)?cdn\.ampproject\.org$/,
        localhostRegex: /^https?:\/\/localhost(:\d+)?$/,
        errorReporting: 'https://amp-error-reporting.appspot.com/r',
        localDev: !1,
        trustedViewerHosts: [
          /(^|\.)google\.(com?|[a-z]{2}|com?\.[a-z]{2}|cat)$/,
          /(^|\.)gmail\.(com|dev)$/,
        ],
        geoApi: void 0,
      },
```

Which is just what I expected (notice the `geoApi: void 0`, which means the new transformer ran). I'd rather use the `MemberExpression` approach because then we're not tied to the explicit patterns that are currently used, and the file can be freely refactored as long as we use `env` as the variable name.